# Estadística Descriptiva y Pandas

El objetivo del siguiente notebook es, además de repasar conceptos de estadística, que sigas aprendiendo a operar con Numpy y que, además, incorpores Pandas a tu caja de herramientas.

## 1. Estadística Descriptiva

La Estadística Descriptiva nos sirve para comenzar a analizar y entender un conjunto de datos. En el caso de datos numéricos, lo hace obteniendo *valores estadísticos* que, de alguna forma, reemplazan a nuestros datos. Por ejemplo, es muy difícil leer y *entender* la edad de 1000 personas. Pero con un grupo reducido de valores estadísticos (mínimo, máximo, media y desviación estándar, etc.) podemos aproximarnos a ese conjunto de una manera mucho más comprensible. Veamos dos medidas muy importantes:

**Promedio**

Dados $n$ números $x_1,x_2,...,x_n$, el promedio o media es 

$$\overline{x} = \frac{1}{n}\sum_{i=1}^{n} x_i = \frac{x_1 + x_2 + ... + x_n}{n}$$

**Desviación Estándar**

La varianza y la desviación estándar nos dan una idea de cuán "dispersos" están los valores con respecto a su promedio.

$$ Var = \frac{\sum_{i=1}^{n} (x_i -\overline{x})^2}{n - 1}$$

La desviación estándar es la raiz cuadrada de la varianza. En general se usa la letra griega $\sigma$ para representarla o las siglas $SD$:

$$ SD = \sqrt{\frac{\sum_{i=1}^{n} (x_i -\overline{x})^2}{n - 1}}$$

$$ SD = \sqrt{Var}$$


**Comentarios**:
1. Dado un conjunto de números, el promedio suele ser considerado el número más representativo de ese conjunto. Esto no siempre es así. Pensá o googleá por qué.
2. Al conjunto de números $x_1,...,x_n$ los pueden encontrar por el nombre de *población* o *muestra* (¡Ojo que no estamos diciendo que *población* y *muestra* sean lo mismo!).

### Challenge:

Vamos a utilizar de excusa la estadística descriptiva para hacer un desafío de programación:

Dadas la siguiente lista de números, escribir una rutina que calcule su promedio, su varianza y desviación estándar. **Pistas:**
* Probablemente te sea muy útil usar lo que hiciste para ejercicios anteriores.
* Para calcular la varianza y la desviación estándar, usa el resultado que obtuviste al calcular el promedio.

El promedio no es representativo por los outliers

In [2]:
x_s = [1,2,3,1,2,2,3,4,1,2,3,4,1,2,4]
sum = 0
for v in x_s:
    sum = sum + v

promedio = sum / len(x_s)
print(promedio)

2.3333333333333335


In [3]:
sumcuad = 0
for v in x_s:
    sumcuad = sumcuad + (v - promedio)**2
    
desvio = sumcuad / (len(x_s) - 1)
print(desvio)

var = desvio ** (1/2)
print(var)

1.2380952380952384
1.1126972805283737


¿Cómo te fue con el Challenge? Si no pudiste resolverlo, no te preocupes. ¡NumPy tiene funciones ya incorporadas que calcula algunos estadísticos sobre un arreglo!

### 1.2 Estadística con NumPy

Veamos cómo se calculan, en NumPy, el promedio, varianza y desviación estándar sobre un arreglo.

In [4]:
import numpy as np

x_s = np.array([1,2,3,1,2,2,3,4,1,2,3,4,1,2,4])

# Promedio
print(x_s.mean())

# Varianza
print(x_s.var(ddof = 1))

# Desviación estándar
print(x_s.std(ddof = 1))

2.3333333333333335
1.238095238095238
1.1126972805283735


**Para investigar**: ¿qué es el parámetro `ddof` de esa función?¿Qué pasa si no lo usas? Esta pregunta es **difícil** y requiere cierto conocimiento previo. Pero intenta, de todas formas, averiguarlo.

NumPy también puede calcular percentilos (¡googlear!), cuantilos, mínimos y máximos:

ddof son los grados de libertad que tiene el calculo que se esta haciendo

In [6]:
print(np.percentile(x_s,75))
print(np.quantile(x_s,0.5))
print(np.min(x_s))
print(np.max(x_s))

3.0
2.0
1
4


**Para investigar**: ¿Cuál es la diferencia entre `np.percentile()` y `np.quantile()`?¿Cómo obtendrías los cuartiles a partir de ellos?

### 1.3 Generación de muestras al azar

Una cosa sumamente útil que podemos hacer con NumPy es generar muestras al azar. Esto no permite simular situaciones. Por ejemplo, las tiradas de dados que aparecen en el GIF de la bitácora. Estas funciones las encontramos dentro del paquete `random` de NumPy, cuya documentación pueden encontrar [aquí](https://docs.scipy.org/doc/numpy-1.15.0/reference/routines.random.html). Veamos cómo lo podemos hacer:

In [5]:
muestras_dado = np.random.randint(7, size = 15)
print(muestras_dado)

### También se puede
muestras_dado = np.random.choice([1,2,3,4,5,6], size = 15)
print(muestras_dado)

[4 1 6 2 0 6 1 2 4 0 1 4 5 2 2]
[2 5 5 5 5 6 3 2 2 2 4 4 4 1 4]


### Ejercitación

**Ejercicio 1:** ¿Cuál será el promedio de los valores obtenidos al tirar muchas veces un dado?¿Te animás a averiguar - o calcular - cuánto *debería dar* antes de hacerlo? Vamos a tratar de responder esta pregunta **simulando** un dado. Para ello:
* Obtener muestras al azar de un dado usando lo que vimos anteriormente.
* Calcular su promedio y desviación estándar.

¿A partir de qué cantidad de muestras el promedio se "estabiliza"?

In [6]:
muestras_dado = np.random.choice([1,2,3,4,5,6], size = 15)

print(muestras_dado)

# Promedio
print(muestras_dado.mean())

# Varianza
print(muestras_dado.var(ddof = 1))

# Desviación estándar
print(muestras_dado.std(ddof = 1))

[4 2 2 5 3 1 3 4 6 1 6 3 5 4 3]
3.466666666666667
2.5523809523809526
1.5976172734359606


**Ejercicio 2:** Simular un dado cargado para favorecer un valor de su elección. Por ejemplo, el seis. Para ello, consultar la ayuda de la función `np.random.choice`. ¿Cómo se modifica el promedio y la desviación estándar?

In [7]:
muestras_dado = np.random.choice([1,2,3,4,5,6], p=[0.1, 0.1, 0.1, 0.1, 0.1, 0.5], size = 15)

print(muestras_dado)

# Promedio
print(muestras_dado.mean())

# Varianza
print(muestras_dado.var(ddof = 1))

# Desviación estándar
print(muestras_dado.std(ddof = 1))

[2 6 2 2 3 6 1 1 6 6 6 5 6 6 3]
4.066666666666666
4.352380952380953
2.0862360730226466


## 2. Pandas 

Pandas es la librería más conocida de Python para manipular y analizar datos. Está montada sobre NumPy, por lo cual muchas funcionalidades son similares. Utilizaremos Pandas para trabajar con datasets estructurados (y bueno, ¡bastante más!). 

Así como NumPy nos proveé de los *arreglos* y con ellos accedemos a muchas nuevas funcionalidades, Pandas nos provee de los *Data Frames* y las *Series*. Por lejos, el objeto más utilizados es el primero, los Data Frames. 


En esta sección empezaremos a:

1. Familiarizarnos con los Data Frames de Pandas, manipular sus funciones básicas y entender la lógica de las mismas (¡para después googlearlas!).
2. Empezar a trabajar con Datasets.

**¡Manos a la obra!**

### 2.1 Primeros pasos

Importamos la librería.

In [8]:
import pandas as pd

Vamos a crear nuestro propio dataset. Es decir, agarrar a mano los datos poblacionales de http://www.ign.gob.ar/nuestrasactividades/geografia/datosargentina/divisionpolitica y guardarlos en una variable `data_dic`.¿Qué tipo de variable es, desde el punto de vista de la programación?

**Nota**: la población está en número de habitantes y la superficie en km2.

In [9]:
data_dic = {"Jurisdiccion":["CABA","Buenos Aires","Catamarca","Chaco","Chubut","Córdoba","Jujuy","Mendoza","Misiones","Río Negro","Santa Cruz",
                           "Santa Fe"],"Poblacion":[2890151,15625084,367828,1055259,509108,3308876,673307,1738929,
                                                   1101593,638645,273964,3194537],"Superficie":
           [200,307521,102606,99633,509108,165321,53219,148827,29801,203013,243943,133007]}

Así como podemos crear arreglos a partir de listas, podemos crear Data Frames a partir de diccionarios.

In [10]:
# Creamos el DataFrame
data_pandas = pd.DataFrame(data_dic)
data_pandas

Jurisdiccion  Poblacion  Superficie
0           CABA    2890151         200
1   Buenos Aires   15625084      307521
2      Catamarca     367828      102606
3          Chaco    1055259       99633
4         Chubut     509108      509108
5        Córdoba    3308876      165321
6          Jujuy     673307       53219
7        Mendoza    1738929      148827
8       Misiones    1101593       29801
9      Río Negro     638645      203013
10    Santa Cruz     273964      243943
11      Santa Fe    3194537      133007

**Ejercicio 1:** investigar las funciones que se implementan en la próxima celda. ¿Qué hacen? ¿Para qué piensan que pueden ser útiles?

In [12]:
data_pandas.head(2)


Jurisdiccion  Poblacion  Superficie
0          CABA    2890151         200
1  Buenos Aires   15625084      307521

In [13]:
data_pandas.tail()


Jurisdiccion  Poblacion  Superficie
7       Mendoza    1738929      148827
8      Misiones    1101593       29801
9     Río Negro     638645      203013
10   Santa Cruz     273964      243943
11     Santa Fe    3194537      133007

In [14]:
data_pandas.count()

Jurisdiccion    12
Poblacion       12
Superficie      12
dtype: int64

In [15]:
data_pandas.shape

(12, 3)

**Ejercicio 2:** agregar al Dataset la información correspondiente a alguna jurisdicción faltante. Recuerden que, al tratarse de una nueva instancia, corresponde a una fila. Pista: googlear "add row to pandas dataframe" o similar. No hay una única forma de hacerlo.

In [18]:
data_dic2 = {"Jurisdiccion":["Tierra del Fuego, Antártida e Islas del Atlántico Sur", "Tucumán", "Entre Ríos"],
            "Poblacion": [127.205, 1448188, 1235994],
            "Superficie": [1002445, 22524, 78781]}

In [20]:
data_pandas2 = pd.DataFrame(data_dic2)
data_pandas2

Jurisdiccion    Poblacion  Superficie
0  Tierra del Fuego, Antártida e Islas del Atlánt...      127.205     1002445
1                                            Tucumán  1448188.000       22524
2                                         Entre Ríos  1235994.000       78781

In [21]:
data_pandas.append(data_pandas2) 

Jurisdiccion     Poblacion  \
0                                                CABA  2.890151e+06   
1                                        Buenos Aires  1.562508e+07   
2                                           Catamarca  3.678280e+05   
3                                               Chaco  1.055259e+06   
4                                              Chubut  5.091080e+05   
5                                             Córdoba  3.308876e+06   
6                                               Jujuy  6.733070e+05   
7                                             Mendoza  1.738929e+06   
8                                            Misiones  1.101593e+06   
9                                           Río Negro  6.386450e+05   
10                                         Santa Cruz  2.739640e+05   
11                                           Santa Fe  3.194537e+06   
0   Tierra del Fuego, Antártida e Islas del Atlánt...  1.272050e+02   
1                                             Tucumán  1.448188e+06   
2                                          Entre Ríos  1.235994e+06   

    Superficie  
0          200  
1       307521  
2       102606  
3        99633  
4       509108  
5       165321  
6        53219  
7       148827  
8        29801  
9       203013  
10      243943  
11      133007  
0      1002445  
1        22524  
2        78781

**Ejercicio 3:** Investigar las funciones columns e index. ¿Qué hacen? ¿Qué tipo de dato es su salida?¿A qué tipo de dato conocido se parecen?

In [22]:
data_pandas.columns

Index(['Jurisdiccion', 'Poblacion', 'Superficie'], dtype='object')

In [23]:
data_pandas.index

RangeIndex(start=0, stop=12, step=1)

**Ejercicio 4:** ¿Qué hacen las siguientes operaciones?

In [24]:
data_pandas['Jurisdiccion']

0             CABA
1     Buenos Aires
2        Catamarca
3            Chaco
4           Chubut
5          Córdoba
6            Jujuy
7          Mendoza
8         Misiones
9        Río Negro
10      Santa Cruz
11        Santa Fe
Name: Jurisdiccion, dtype: object

In [25]:
data_pandas[['Jurisdiccion','Poblacion']]

Jurisdiccion  Poblacion
0           CABA    2890151
1   Buenos Aires   15625084
2      Catamarca     367828
3          Chaco    1055259
4         Chubut     509108
5        Córdoba    3308876
6          Jujuy     673307
7        Mendoza    1738929
8       Misiones    1101593
9      Río Negro     638645
10    Santa Cruz     273964
11      Santa Fe    3194537

In [26]:
data_pandas.Jurisdiccion

0             CABA
1     Buenos Aires
2        Catamarca
3            Chaco
4           Chubut
5          Córdoba
6            Jujuy
7          Mendoza
8         Misiones
9        Río Negro
10      Santa Cruz
11        Santa Fe
Name: Jurisdiccion, dtype: object

In [27]:
'Poblacion' in data_pandas

True

**Ejercicio 5:** Agregar una columna al dataframe que corresponda a la densidad de cada jurisdicción. Usar la información que **ya está** en el dataset.

In [28]:
data_pandas['Densidad'] = data_pandas.apply(lambda row: row.Poblacion / row.Superficie, axis=1)

In [29]:
data_pandas

Jurisdiccion  Poblacion  Superficie      Densidad
0           CABA    2890151         200  14450.755000
1   Buenos Aires   15625084      307521     50.809811
2      Catamarca     367828      102606      3.584859
3          Chaco    1055259       99633     10.591461
4         Chubut     509108      509108      1.000000
5        Córdoba    3308876      165321     20.014856
6          Jujuy     673307       53219     12.651628
7        Mendoza    1738929      148827     11.684231
8       Misiones    1101593       29801     36.964968
9      Río Negro     638645      203013      3.145833
10    Santa Cruz     273964      243943      1.123066
11      Santa Fe    3194537      133007     24.017811

### 2.2 Filtrado por máscara.

Lo que veremos a continuación es **muy importante**, ya que es una operación que haremos muchas veces. Su implementación es muy parecida tanto en NumPy como en Pandas, por lo que veremos cómo hacerlo primero en NumPy luego en Pandas.

Supongamos que hacemos 50 tiradas de un dado, como hicimos en la sección anterior, pero queremos seleccionar solamente aquellas tiradas que fueron menores que cuatro. ¿Cómo podemos hacerlo?

In [30]:
muestras_dado = np.random.randint(1,7, size = 50)
print(muestras_dado)

[3 3 4 4 5 2 6 1 3 3 5 3 4 5 3 5 2 6 4 5 2 1 2 1 1 6 1 6 5 1 4 2 3 4 6 3 3
 4 6 6 3 3 4 1 4 6 2 2 2 5]


Lo que podemos hacer es crear una máscara:

In [31]:
mascara = muestras_dado < 4
print(mascara)
print(type(mascara))

[ True  True False False False  True False  True  True  True False  True
 False False  True False  True False False False  True  True  True  True
  True False  True False False  True False  True  True False False  True
  True False False False  True  True False  True False False  True  True
  True False]
<class 'numpy.ndarray'>


Notar que `mascara` es un arreglo de booleanos, con `True` en los valores que cumplen la condición y `False` donde no. Una vez que creamos la máscara, podemos usarla para seleccionar de nuestro arreglo aquellos elementos que queríamos:

In [32]:
print(muestras_dado[mascara])

[3 3 2 1 3 3 3 3 2 2 1 2 1 1 1 1 2 3 3 3 3 3 1 2 2 2]


Notar que con `mascara.sum()` podemos contar cuántas veces se cumple la condición que pedimos.

In [33]:
print(mascara.sum())

26


A veces, podemos hacerlo en una sola línea. Supongamos que queremos aquellas tiradas donde salió seis:

In [34]:
print(muestras_dado[muestras_dado == 6])

[6 6 6 6 6 6 6 6]


**En Pandas**

Supongamos que queremos seleccionar aquellas jurisdicciones cuya población sea mayor a un millón de habitantes. Podemos hacerlo de la siguiente forma:

In [35]:
data_pandas[data_pandas.Poblacion > 1000000]

Jurisdiccion  Poblacion  Superficie      Densidad
0           CABA    2890151         200  14450.755000
1   Buenos Aires   15625084      307521     50.809811
3          Chaco    1055259       99633     10.591461
5        Córdoba    3308876      165321     20.014856
7        Mendoza    1738929      148827     11.684231
8       Misiones    1101593       29801     36.964968
11      Santa Fe    3194537      133007     24.017811

¿Y si queremos seleccionar aquellas jurisdicciones cuya población sea mayor a un millón de habitantes **y** su superficie menor a cien mil km2?

In [36]:
mascara = np.logical_and(data_pandas.Poblacion > 1000000, data_pandas.Superficie < 100000)
data_pandas[mascara]

Jurisdiccion  Poblacion  Superficie      Densidad
0         CABA    2890151         200  14450.755000
3        Chaco    1055259       99633     10.591461
8     Misiones    1101593       29801     36.964968

In [37]:
### Es equivalente
data_pandas[(data_pandas.Poblacion > 1000000) & (data_pandas.Superficie < 100000)]

Jurisdiccion  Poblacion  Superficie      Densidad
0         CABA    2890151         200  14450.755000
3        Chaco    1055259       99633     10.591461
8     Misiones    1101593       29801     36.964968

**Ejercicio:** seleccionar aquellas jurisdicciones cuya población sea menor a 500 mil habitantes **o** su superficie mayor a cien mil km2.

In [38]:
mascara = np.logical_or(data_pandas.Poblacion < 500000, data_pandas.Superficie > 100000)
data_pandas[mascara]

Jurisdiccion  Poblacion  Superficie   Densidad
1   Buenos Aires   15625084      307521  50.809811
2      Catamarca     367828      102606   3.584859
4         Chubut     509108      509108   1.000000
5        Córdoba    3308876      165321  20.014856
7        Mendoza    1738929      148827  11.684231
9      Río Negro     638645      203013   3.145833
10    Santa Cruz     273964      243943   1.123066
11      Santa Fe    3194537      133007  24.017811

## 2.3 Iris dataset

¿Pero siempre vamos a tener que crear un diccionario y luego pasarlo a un Data Frame? Evidentemente, esta opción no parece muy cómoda, en particular para conjuntos de datos de gran volumen. Veamos cómo trabajamos con un conjunto de datos preexistente.


**Aviso**: Esta sección es, en realidad, un ejercicio. Para hacerlo, debes ir googleando y consultando la documentación que consideres apropiada. Obviamente, también puedes consultar a tu mentor/a.

Vamos a trabajar con el Iris Dataset, probablemente uno de los conjuntos de datos más famosos, ya que muchos ejemplos se realizan con él. Es un dataset sencillo pero ilustrativo.


1. Abrir con Pandas el archivo 'DS_Bitácora_04_iris.csv' (¿Qué tipo de archivo es?) e imprimir sus primeros cinco elementos. Pista: `pd.read...()`.

In [41]:
df = pd.read_csv('DS_Bitácora_04_Iris.csv')  

In [42]:
df

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
..   ...            ...           ...            ...           ...   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
0       Iris-setosa  
1       Iris-setosa  
2       Iris-setosa  
3       Iris-setosa  
4       Iris-setosa  
..              ...  
145  Iris-virginica  
146  Iris-virginica  
147  Iris-virginica  
148  Iris-virginica  
149  Iris-virginica  

[150 rows x 6 columns]

In [43]:
df.head(5)

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

2. ¿Cuántas columnas (features) tiene?¿Cuáles son sus nombres?¿Y cuántas filas (instancias)? Pistas: `shape`, `columns`.

In [44]:
df.columns

Index(['Id', 'SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm',
       'Species'],
      dtype='object')

In [47]:
df.shape

(150, 6)

# Iris dataset
Cuenta con 150 registros 
Tiene 6 columnas


3. Obtener el valor medio y desviación estándar de cada columna. ¿Hay alguna función de Pandas que nos dé aún más estadísticos? Pistas: `describe`.

In [48]:
df.describe()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count  150.000000     150.000000    150.000000     150.000000    150.000000
mean    75.500000       5.843333      3.054000       3.758667      1.198667
std     43.445368       0.828066      0.433594       1.764420      0.763161
min      1.000000       4.300000      2.000000       1.000000      0.100000
25%     38.250000       5.100000      2.800000       1.600000      0.300000
50%     75.500000       5.800000      3.000000       4.350000      1.300000
75%    112.750000       6.400000      3.300000       5.100000      1.800000
max    150.000000       7.900000      4.400000       6.900000      2.500000

4. ¿Creen que todas las columnas tienen información? *Tirar* la columna que crean que está demás. Dependiendo de la función que uses - hay más de una opción -, tal vez tengas que prestar **mucha** atención al argumento `inplace`. Pista: `drop`, `del`.

In [50]:
df.isna()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
0    False          False         False          False         False    False
1    False          False         False          False         False    False
2    False          False         False          False         False    False
3    False          False         False          False         False    False
4    False          False         False          False         False    False
..     ...            ...           ...            ...           ...      ...
145  False          False         False          False         False    False
146  False          False         False          False         False    False
147  False          False         False          False         False    False
148  False          False         False          False         False    False
149  False          False         False          False         False    False

[150 rows x 6 columns]

In [51]:
df.drop(['Id'], axis=1, inplace=True)

In [52]:
df

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm         Species
0              5.1           3.5            1.4           0.2     Iris-setosa
1              4.9           3.0            1.4           0.2     Iris-setosa
2              4.7           3.2            1.3           0.2     Iris-setosa
3              4.6           3.1            1.5           0.2     Iris-setosa
4              5.0           3.6            1.4           0.2     Iris-setosa
..             ...           ...            ...           ...             ...
145            6.7           3.0            5.2           2.3  Iris-virginica
146            6.3           2.5            5.0           1.9  Iris-virginica
147            6.5           3.0            5.2           2.0  Iris-virginica
148            6.2           3.4            5.4           2.3  Iris-virginica
149            5.9           3.0            5.1           1.8  Iris-virginica

[150 rows x 5 columns]

5. ¿Para qué sirven `loc` e `iloc`? Crea algunos ejemplos.

In [54]:
df.loc[5, 'Species']

'Iris-setosa'

In [55]:
df.loc[:5, ['SepalLengthCm', 'Species']]

SepalLengthCm      Species
0            5.1  Iris-setosa
1            4.9  Iris-setosa
2            4.7  Iris-setosa
3            4.6  Iris-setosa
4            5.0  Iris-setosa
5            5.4  Iris-setosa

In [58]:
df[df.PetalLengthCm < 3].loc[:5, ['SepalLengthCm', 'Species']]

SepalLengthCm      Species
0            5.1  Iris-setosa
1            4.9  Iris-setosa
2            4.7  Iris-setosa
3            4.6  Iris-setosa
4            5.0  Iris-setosa
5            5.4  Iris-setosa

In [59]:
df.iloc[:3, :3]

SepalLengthCm  SepalWidthCm  PetalLengthCm
0            5.1           3.5            1.4
1            4.9           3.0            1.4
2            4.7           3.2            1.3